### ISSUE TITLE
FM / FFM / Wide&Deep / DeepFFM

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9307]
FM / FFM / Wide&amp;Deep / DeepFFM<br>
<br>
https://ift.tt/REKoXBY<br>
<br>
贪钱算法还我头发<br>
<br>

<div><div><div><div><article><h1>FM</h1>
<p>LR + 特征交叉项：进行特征组合增强模型泛化能力。</p>
<p>$$ y=w_{0}+\sum_{i=1}^{n}{w_{i}x_{i}}+\sum_{i=1}^{n}{\sum_{i+1}^{n}{&lt;v_{i},v_{j}&gt;x_{i}x_{j}}} $$ </p>
<p>式中$v_i$为特征的embedding向量</p>
<h1>FFM</h1>
<p>FM + Field：将特征按照事先规则分多个场。每个特征将被映射为多个隐向量，每个隐向量对应一个场。当两个特征组合时，用对方对应的场对应的隐向量做内积。FM可看做只有一个场的FFM。</p>
<p>$$ y=w_{0}+\sum_{i=1}^{n}{w_{i}x_{i}}+\sum_{i=1}^{n}{\sum_{i+1}^{n}{&lt;v_{i,f_j}, v_{j,f_i}&gt;x_{i}x_{j}}} $$ </p>
<h1>Wide &amp; Deep</h1>
<p>Wide + Deep：由浅层（或单层）的Wide部分神经网络和深层的Deep部分多层神经网络组成，输出层采用softmax或logistics regression综合Wide和Deep部分的输出。Wide部分有利于增强模型的“记忆能力”，Deep部分有利于增强模型的“泛化能力”。</p>
<p>$$ P(Y=1|x)=\sigma(W^T_{wide}[x, \phi (x)] + W^T_{deep}a^{l_f} + b) $$</p>
<center>
<img alt="Wide &amp; Deep" align="center" src="https://xfliu1998.github.io/2022/03/17/7-FM-FFM-Wide&amp;Deep-DeepFFM/Wide&amp;Deep.png"><p>Wide &amp; Deep 网络架构</p>
</center>
<h1>DeepFM</h1>
<p>Deep + FM：用FM做特征间低阶组合，用Deep NN部分做特征间高阶组合，并行方式组合。DeepFM不需要预训练和人工特征工程。</p>
<p>$$\hat y=sigmoid(y_{FM}+y_{DNN})$$</p>
<center>
<img alt="DeepFM" align="center" src="https://xfliu1998.github.io/2022/03/17/7-FM-FFM-Wide&amp;Deep-DeepFFM/DeepFM.png"><p>DeepFM 网络架构</p>
</center>
<p><code>tf.nn.embedding_lookup_sparse()</code> 多值离散特征的embedding处理</p>
<p>Deep + Cross: 深度神经网络+应用显式特征交叉。特征交叉部分公式为：</p>
<p>$$x_{l+1}=x_0x^T_lw_l+b_l+x_l=f(x_l, w_l, b_l)+x_l$$</p>
<center>
<img alt="Deep&amp;Cross" align="center" src="https://xfliu1998.github.io/2022/03/17/7-FM-FFM-Wide&amp;Deep-DeepFFM/Deep&amp;Cross.png"><p>Deep &amp; Cross 网络架构</p>
</center>
<h1>面试问题</h1>
<ol>
<li>DeepFM和Wide&amp;Deep的deep部分有什么不同？<br>原生的wide&amp;deep里面的deep输入是onehot，数值特征和标签特征都有；DeepFM中的deep是连续数值的embedding输入，且与FM的embedding共享，训练更快更准确。</li>
</ol>
<h1>参考资料</h1>
</article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via 一直进步 做喜欢的 <br>
<br>
May 2, 2023 at 08:34AM
